guide/src/class/protocols.md

```diff
@@ -147,6 +141,108 @@ Other:
+
+    Implements membership test operators.
+    Should return true if `item` is in `self`, false otherwise.
+    For objects that don’t define `__contains__()`, the membership test simply
+    traverses the sequence until it finds a match.
+
+  * `fn __concat__(&self, other: impl FromPyObject) -> PyResult<impl ToPyObject>`
+
+    Concatenates two sequences.
+    Used by the `+` operator, after trying the numeric addition via
+    the `PyNumberProtocol` trait method.
+
+  * `fn __repeat__(&self, count: isize) -> PyResult<impl ToPyObject>`
+
+    Repeats the sequence `count` times.
+    Used by the `*` operator, after trying the numeric multiplication via
+    the `PyNumberProtocol` trait method.
+
+  * `fn __inplace_concat__(&mut self, other: impl FromPyObject) -> PyResult<Self>`
+
+    Concatenates two sequences in place. Returns the modified first operand.
+    Used by the `+=` operator, after trying the numeric in place addition via
+    the `PyNumberProtocol` trait method.
+
+  * `fn __inplace_repeat__(&mut self, count: isize) -> PyResult<Self>`
+
+    Repeats the sequence `count` times in place. Returns the modified first operand.
+    Used by the `*=` operator, after trying the numeric in place multiplication via
+    the `PyNumberProtocol` trait method.
+
+### Emulating mapping containers (such as dictionaries)
+
+The [`PyMappingProtocol`] trait allows to emulate
+[mapping container types](https://docs.python.org/3/reference/datamodel.html#emulating-container-types).
+
+For a mapping, the keys may be Python objects of arbitrary type.
+
+  * `fn __len__(&self) -> PyResult<usize>`
+
+    Implements the built-in function `len()` for the mapping.
+
+  * `fn __getitem__(&self, key: impl FromPyObject) -> PyResult<impl ToPyObject>`
+
+    Implements evaluation of the `self[key]` element.
+    If `key` is of an inappropriate type, `TypeError` may be raised;
+    if `key` is missing (not in the container), `KeyError` should be raised.
+
+  * `fn __setitem__(&mut self, key: impl FromPyObject, value: impl FromPyObject) -> PyResult<()>`
+
+    Implements assignment to the `self[key]` element or insertion of a new `key`
+    mapping to `value`.
+    Should only be implemented if the mapping support changes to the values for keys,
+    or if new keys can be added.
+    The same exceptions should be raised for improper key values as
+    for the `__getitem__()` method.
+
+  * `fn __delitem__(&mut self, key: impl FromPyObject) -> PyResult<()>`
+
+    Implements deletion of the `self[key]` element.
+    Should only be implemented if the mapping supports removal of keys.
+    The same exceptions should be raised for improper key values as
+    for the `__getitem__()` method.
+
+  * `fn __reversed__(&self) -> PyResult<impl ToPyObject>`
+
+    Called (if present) by the `reversed()` built-in to implement reverse iteration.
+    It should return a new iterator object that iterates over all the objects in
+    the container in reverse order.
```

Comment:
Hmm we should probably move `__reversed__` to the `PyIterProtocol` trait? (I can do it in a separate PR.)

Reply:
Well, `__reversed__()` is a standalone method, not a [`PyMappingMethods`](https://docs.python.org/3/c-api/typeobj.html#c.PyMappingMethods) slot, so there is no particular reason it should be a part of `PyMappingProtocol`.
However, implementing `__reversed__()` for sequences is not very useful, because [the default implementation](https://docs.python.org/3/reference/datamodel.html#object.__reversed__) seems sufficient for most cases.

I personally do not like the idea of adding anything to `PyIterProtocol` though. It's already semantically overloaded as it encompasses both `IntoIterator` and `Iterator` Rust traits. But then the user is expected to implement a different subset of methods for each of the intended uses (as an iterator or as an iterable).

Currently, we also have to write mandatory boilerplate code like
```rust
fn __iter__(slf: PyRef<Self>) -> PyRef<Self> {
    slf
}
```
for every custom iterator class.

Most Rust programmers would expect a blanket implementation like
```rust
impl<I> IntoIterator for I where
    I: Iterator, 
```
from the stdlib to exist instead.

I'm sorry about the long rant, but `PyIterProtocol` was the biggest road bump for me when working with PyO3.
I really wish it was two distinct traits, like for example `PyIterProtocol` and `PyIntoIterProtocol`, so that anyone who knows how Rust iterators work could easily implement Python iterators without scratching their head about "why do I need to return `PyRef<Self>` instead `Self` here?"

Reply:
Thanks, I completely agree that the current situation is confusing in many ways. I wonder if alternatively here I should remove `__reversed__` from `#[pyproto]` completely. In my opinion `#[pymethods]` are very easy to learn, and implementing `__reversed__` in `#[pymethods]` will already work.

Reply:
I like this idea very much. There are other non-slotted methods that are included in the protocol traits, like `__bytes__()`. And then there are C function slots that have no corresponding `PyObjectProtocol` method like [`tp_call`](https://docs.python.org/3/c-api/typeobj.html#c.PyTypeObject.tp_call) aka [`__call__()`](https://docs.python.org/3/reference/datamodel.html#object.__call__).

I'm all for implementing true non-slotted methods as `#[pymethods]`. This way we can avoid surprises like https://github.com/PyO3/pyo3/issues/1465 in the future.